zerver/tests/test_markdown_thumbnail.py

```diff
@@ -0,0 +1,309 @@
+                f'<div class="message_inline_image"><a href="/user_uploads/{path_ids[1]}" title="{image_names[1]}">'
+                f'<img src="/user_uploads/thumbnail/{path_ids[1]}/840x560.webp"></a></div>'
+                "<p>Another screenshot<br>\n"
+                f'<a href="/user_uploads/{path_ids[2]}">{image_names[2]}</a></p>\n'
+                f'<div class="message_inline_image"><a href="/user_uploads/{path_ids[2]}" title="{image_names[2]}">'
+                f'<img src="/user_uploads/thumbnail/{path_ids[2]}/840x560.webp"></a></div>'
+            ),
+        )
+
+    def test_thumbnail_code_block(self) -> None:
+        url = "http://example.com/image.png"
+        path_id = self.upload_and_thumbnail_image("img.png")
+        # We have a path_id of an image in the message content, so we
+        # will prefetch the thumbnail metadata -- but not insert it.
+
+        sender_user_profile = self.example_user("othello")
+        msg = Message(
+            sender=sender_user_profile,
+            sending_client=get_client("test"),
+            realm=sender_user_profile.realm,
+        )
+        converted = render_message_markdown(msg, f"{url}\n```\n/user_uploads/{path_id}\n```")
+        self.assertEqual(
+            converted.rendered_content,
+            (
+                f'<div class="message_inline_image"><a href="{url}"><img src="{get_camo_url(url)}"></a></div>'
+                f'<div class="codehilite"><pre><span></span><code>/user_uploads/{path_id}\n'
+                "</code></pre></div>"
+            ),
+        )
+
+    def test_thumbnail_after_send(self) -> None:
+        with self.captureOnCommitCallbacks(execute=True):
+            path_id = self.upload_image("img.png")
+            content = f"[image](/user_uploads/{path_id})"
+            expected = (
+                f'<p><a href="/user_uploads/{path_id}">image</a></p>\n'
+                f'<div class="message_inline_image"><a href="/user_uploads/{path_id}" title="image">'
+                '<img class="image-loading-placeholder" src="/static/images/loading/loader-black.svg"></a></div>'
+            )
+
+            message_id = self.send_message_content(content)
+            self.assert_message_content_is(message_id, expected)
+
+            # Exit the block and run thumbnailing
+        expected = (
+            f'<p><a href="/user_uploads/{path_id}">image</a></p>\n'
+            f'<div class="message_inline_image"><a href="/user_uploads/{path_id}" title="image">'
+            f'<img src="/user_uploads/thumbnail/{path_id}/840x560.webp"></a></div>'
+        )
+        self.assert_message_content_is(message_id, expected)
+
+    def test_thumbnail_repeated(self) -> None:
+        # We currently have no way to generate a thumbnailing event
+        # for the worker except during upload, meaning that we will
+        # never repeat a ImageAttachment thumbnailing.  However, the
+        # code supports it, so test it.
```

Comment:
I guess one can imagine migration/defect-correction use cases for this capability.

Reply:
And any change to if we try to thumbnail in Django, as we've discussed elsewhere, would hit this.

Reply:
Hmm, I would have thought the row lock would mean doing it in uwsgi workers wouldn't hit this?